src/lib/FreeC/LiftedIR/Syntax/Expr.hs

```diff
@@ -0,0 +1,116 @@
+-- | This module contains the definition of expressions for our lifted
+--   intermediate language.
+
+module FreeC.LiftedIR.Syntax.Expr where
+
+import           FreeC.IR.SrcSpan               ( SrcSpan )
+
+import           FreeC.LiftedIR.Effect          ( Effect )
+import           FreeC.LiftedIR.Syntax.Name
+import           FreeC.LiftedIR.Syntax.Type
+
+-- | An expression.
+data Expr
+  = -- | A constructor.
+    Con { exprSrcSpan :: SrcSpan
+        , exprConName :: ConName
+        , exprType    :: Type
+        }
+
+  | -- | A function or local variable.
+    Var { exprSrcSpan :: SrcSpan
+        , exprVarName :: VarName
+        , exprType    :: Type
+        }
+
+  | -- | Function or constructor application.
+    App { exprSrcSpan     :: SrcSpan
+        , exprAppFunc     :: Expr
+        , exprAppTypeArgs :: [Type]   -- ^ Visible type applications.
+        , exprEffects     :: [Effect] -- ^ Effect set.
```

Comment:
Is the effect set of an application supposed to be the union of the effects of its arguments? Is there a benefit in having the effect set information here instead of as part of the `Free` type constructor?

Reply:
The `exprEffects` field models the type class instance arguments that have to be passed to effectful functions in Coq and Agda. For example, `head` is defined as follows in Coq.

```haskell
Definition head (Shape : Type) (Pos : Shape -> Type) (P : Partial Shape Pos) {a : Type} … .
```

The `Partial` instance is needed such that `head` can use the `undefined` and `error` functions on its right-hand side.
If `head` is applied, the `Partial` instance `P` has to be passed explicitly.

```haskell
@head Shape Pos P a xs
```

The `exprEffects` corresponds exactly to the argument `P` in this case (comparred to the current implementation in Coq, the order of `exprAppTypeArgs` and `exprEffects` has been swapped in lifted IR such that the effects such `State s` could depend on type arguments in the future).
If there were multiple effects, multiple type class instances would have to be passed. `exprEffects` must be a list instead of a set since the order in which the effects have to be passed around is fixed by the function definition.

In this definition, the effect set of a function declaration is the union of all effect sets of functions that are used on the right-hand side. The effect set of local variables is always empty. For example, even if `xs` contains computations with further effects, they do not belong to `head`'s effect set. An application like `head [1 ? 2]` in Curry, for example, requires the non-determinism effect to be part of the effect set of the caller. However, `head` does not have to know that `Shape` and `Pos` provide a mechanism for non-determinism. It is only interested in the partiality effect that is provided by the same `Shape` and `Pos` in addition to non-determinism.

Reply:
> In this definition, the effect set of a function declaration is the union of all effect sets of functions that are used on the right-hand side. 

Since the effect set is not part of the function's signature, is it stored in the environment or recomputed each time?

Reply:
It is stored in the environment. There is an [`entryIsPartial`](https://github.com/FreeProving/free-compiler/blob/1ca3159363e039f60c2b5cedb10ff5b7938d7f90/src/lib/FreeC/Environment/Entry.hs#L94) flag for `FuncEntry`s which is set by the [`partialityAnalysisPass`](https://github.com/FreeProving/free-compiler/blob/1ca3159363e039f60c2b5cedb10ff5b7938d7f90/src/lib/FreeC/Pass/PartialityAnalysisPass.hs). In the future the flag would be replaced by a list of `Effect`s and the pass would be generalized to arbitrary effects.